src/middlewared/middlewared/plugins/idmap.py

```diff
@@ -0,0 +1,1120 @@
+        """
+        verrors = ValidationErrors()
+        verrors.add_child('idmap_domain_create', await self._validate(data))
+        if verrors:
+            raise verrors
+
+        data["id"] = await self.middleware.call(
+            "datastore.insert", self._config.datastore, data,
+            {
+                "prefix": self._config.datastore_prefix
+            },
+        )
+        return await self._get_instance(data['id'])
+
+    @accepts(
+        Int('id', required=True),
+        Patch(
+            "idmap_domain_create",
+            "idmap_domain_update",
+            ("attr", {"update": True})
+        )
+    )
+    async def do_update(self, id, data):
+        """
+        Update a domain by id.
+        """
+        old = await self._get_instance(id)
+        new = old.copy()
+        new.update(data)
+        verrors = ValidationErrors()
+        verrors.add_child('idmap_domain_update', await self._validate(new))
+
+        if verrors:
+            raise verrors
+
+        await self.middleware.call(
+            'datastore.update',
+            self._config.datastore,
+            id,
+            new,
+            {'prefix': self._config.datastore_prefix}
+        )
+        return await self._get_instance(id)
+
+    @accepts(Int('id'))
+    async def do_delete(self, id):
+        """
+        Delete a domain by id. Deletion of default system domains is not permitted.
+        """
+        if id <= 5:
+            entry = await self._get_instance(id)
+            raise CallError(f'Deleting system idmap domain [{entry["name"]}] is not permitted.', errno.EPERM)
+        await self.middleware.call("datastore.delete", self._config.datastore, id)
+
+    @private
+    async def _validate(self, data):
+        verrors = ValidationErrors()
+        if data['id'] <= dstype['DS_TYPE_DEFAULT_DOMAIN'].value:
+            verrors.add(f'Modifying system idmap domain [{data["name"]}] is not permitted.')
+        return verrors
+
+
+class IdmapDomainBackendService(CRUDService):
+    class Config:
+        datastore = 'directoryservice.idmap_domaintobackend'
+        datastore_prefix = 'idmap_dtb_'
+        namespace = 'idmap.domaintobackend'
+
+    @accepts(
+        Dict(
+            'idmap_domaintobackend_create',
+            Dict(
+                'domain',
+                Int('id'),
+                Str('idmap_domain_name'),
+                Str('idmap_domain_dns_domain_name'),
+            ),
+            Str('idmap_backend', enum=['ad', 'autorid', 'fruit', 'ldap', 'nss', 'rfc2307', 'rid', 'script', 'tdb', 'tdb2']),
+            register=True
+        )
+    )
+    async def do_create(self, data):
+        """
+        Set an idmap backend for a domain.
+        :domain: - dictionary containing domain information. Has one-to-one relationship with idmap_domain entries. id value
+        in dictionary is not required.
+        :idmap_backed: - type of idmap backend to use for the domain.
```

Comment:
We should check how these look in /api/docs/

Noting the attributes with `:attribute:` would probably have no effect on the markup and they will all be concatenated in the same line unless they is a double line break.